README.md
# metapredict

[![check-deploy](https://github.com/hugheylab/metapredict/workflows/check-deploy/badge.svg)](https://github.com/hugheylab/metapredict/actions)
[![codecov](https://codecov.io/gh/hugheylab/metapredict/branch/master/graph/badge.svg?token=nRgjANwZ6s)](https://codecov.io/gh/hugheylab/metapredict)
[![Netlify Status](https://api.netlify.com/api/v1/badges/6ff1bdb7-ea07-4c5f-9a52-f27f9eae8554/deploy-status)](https://app.netlify.com/sites/infallible-spence-3f03b0/deploys)
[![drat version](https://raw.githubusercontent.com/hugheylab/drat/gh-pages/badges/metapredict_drat_badge.svg)](https://github.com/hugheylab/drat/tree/gh-pages/src/contrib)

`metapredict` enables meta-analysis of gene expression using the elastic net (i.e., `glmnet`). For technical details, please check out the [paper](https://doi.org/10.1093/nar/gkv229).

**If you just want to process microarray data and don't need to merge the expression data to perform a meta-analysis, then check out [seeker](https://seeker.hugheylab.org), which makes fetching, processing, and normalizing transcriptome data even simpler.**

## Installation

1. Install [`BiocManager`](https://cran.r-project.org/package=BiocManager).

    ```r
    if (!requireNamespace('BiocManager', quietly = TRUE))
      install.packages('BiocManager')
    ```

1. If you use RStudio, go to Tools → Global Options... → Packages → Add... (under Secondary repositories), then enter:

    - Name: hugheylab
    - Url: https://hugheylab.github.io/drat/

    You only have to do this once. Then you can install or update the package by entering:

    ```r
    BiocManager::install('metapredict')
    ```

    Alternatively, you can install or update the package by entering:

    ```r
    BiocManager::install('metapredict', site_repository = 'https://hugheylab.github.io/drat/')
    ```

## Usage

Go through the vignettes to

- [prepare the data for the example meta-analysis](https://metapredict.hugheylab.org/articles/prepare_example.html),
- [run the example meta-analysis](https://metapredict.hugheylab.org/articles/run_example.html), and
- [see how to run your own meta-analysis](https://metapredict.hugheylab.org/articles/guide.html).

For help on using specific functions, consult the [reference documentation](https://metapredict.hugheylab.org/reference/index.html).


NEWS.md
# metapredict 1.1.4
- Fixed a few issues causing test failures.

# metapredict 1.1.3
- Fixed issues caught by lintr.

# metapredict 1.1.2
- Tweaked usage of `data.table` to avoid warnings.

# metapredict 1.1.1
- Revised code to not need `globalVariables()` in order to pass R CMD check without notes.

# metapredict 1.1.0
- Switched to data.table under the hood.

# metapredict 1.0.4
- Updated links in vignette to download data from the Bhattacharjee study.

# metapredict 1.0.3
- Fixed a bug in mapping probes to genes for some platforms that was caused by mixing up characters and integers.
- Enabled support for gzipped series matrix files.

# metapredict 1.0.2
- Vignettes now properly link to each other and to documentation.

# metapredict 1.0.1
- Added `pkgdown` site.
- Updated roxygen-based documentation.
- `installCustomCdfPackages` now uses latest BrainArray version.
- `metapredict` now works with updated `glmnet::predict.glmnet`.


tests/testthat.R
library('testthat')
library('metapredict')

test_check('metapredict')


tests/testthat/setup.R
library('ggplot2')
library('RColorBrewer')
library('data.table')
library('glmnet')
library('withr')

parentFolderPath = 'test_data'
foldidColname = 'study'
className = 'class'
family = 'multinomial'
alpha = 0.9
discoveryStudyNames = c('Bhattacharjee', 'GSE11969', 'GSE29016')
classLevels = c('AD', 'SQ', 'SCLC')

studyMetadataPath = file.path(parentFolderPath, 'study_metadata.csv')
studyMetadata = read.csv(studyMetadataPath, stringsAsFactors = FALSE)

sampleMetadataPath = file.path(parentFolderPath, 'sample_metadata.csv')
sampleMetadata = read.csv(sampleMetadataPath, stringsAsFactors = FALSE)

esetListControl = readRDS(file.path(parentFolderPath, 'esetList.rds'))

ematListControl = readRDS(file.path(parentFolderPath, 'ematList.rds'))

ematDiscoveryControl = readRDS(file.path(parentFolderPath, 'ematDiscovery.rds'))

glmnetArgsControl = readRDS(file.path(parentFolderPath, 'glmnetArgs.rds'))

cvFitListControl = readRDS(file.path(parentFolderPath, 'cvFitList.rds'))
cvFit = cvFitListControl[[1]]
fitResult = cvFit$glmnet.fit
lambda = cvFit$lambda.min

predsListControl = readRDS(file.path(parentFolderPath, 'predsList.rds'))

calcConfusionValidationControl = readRDS(file.path(parentFolderPath, 'calcConfusionValidation.rds'))

calcConfusionValidationEachFalseControl =
  readRDS(file.path(parentFolderPath, 'calcConfusionValidationEachFalse.rds'))

plotCoefficientsControl = readRDS(file.path(parentFolderPath, 'plotCoefficients.rds'))

annoNames = c('study', 'class')
annoLevels = list(discoveryStudyNames, classLevels)
names(annoLevels) = annoNames
annoColors = list(brewer.pal(3, 'Dark2'), brewer.pal(3, 'Paired'))
names(annoColors) = annoNames
names(annoColors[[1]]) = annoLevels[[1]]
names(annoColors[[2]]) = annoLevels[[2]]

plotExpressionHeatmapControl = readRDS(file.path(parentFolderPath, 'plotExpressionHeatmap.rds'))

plotClassProbsValidationControl = readRDS(file.path(parentFolderPath, 'plotClassProbsValidation.rds'))


tests/testthat/test_metapredict_plot.R


test_that('plotCoefficients', {
  plotCoefficientsTest = plotCoefficients(fitResult, lambda, classLevels = classLevels) +
    scale_fill_brewer(type = 'qual', palette = 3) +
    theme(axis.text.y = element_text(size = 8))
  expect_equal(plotCoefficientsTest$data, plotCoefficientsControl$data, check.attributes = FALSE)
})

test_that('plotExpressionHeatmap', {
  plotExpressionHeatmapTest = plotExpressionHeatmap(fitResult, lambda, ematDiscoveryControl, sampleMetadata, annoLevels,
                                                annoColors, classLevels = classLevels, fontsize_row = 7)
  expect_equal(plotExpressionHeatmapTest$data, plotExpressionHeatmapControl$data, check.attributes = FALSE)
})

test_that('plotClassProbsValidation', {
  plotClassProbsValidationTest = plotClassProbsValidation(predsListControl, sampleMetadata, className, classLevels, size = 1.5,
                                                      ggplotArgs = list(scale_color_brewer(type = 'qual', palette = 3)))
  expect_equal(plotClassProbsValidationTest$data, plotClassProbsValidationControl$data, check.attributes = FALSE)
})


R/metapredict_plot.R
#' Plot coefficients for a gene expression meta-analysis.
#'
#' Make a ggplot of the coefficients from a `glmnet` object trained on a gene
#' expression meta-analysis (either logistic or multinomial classification).
#'
#' @param fitResult glmnet object.
#' @param lambda value of lambda to use in `fitResult`.
#' @param classLevels Order of classes in the confusion matrix. If `NA` (default),
#'   then the function uses the order in `fitResult`.
#' @param decreasing logical indicating whether to sort genes by decreasing coefficient.
#' @param geneIdOrder Optional character array of Entrez Gene IDs specifying the order
#'   of genes. If `NA` (default), the order from [makeCoefDt()] is used.
#' @param org Name of package for mapping Entrez Gene IDs to gene symbols,
#'   passed to `data` argument of [annotate::lookUp()].
#'
#' @return A `ggplot` object.
#'
#' @export
plotCoefficients = function(
  fitResult, lambda, classLevels = NA, decreasing = FALSE, geneIdOrder = NA,
  org = 'org.Hs.eg') {

  geneId = NULL
  coefDt = makeCoefDt(fitResult, lambda, decreasing = decreasing, classLevels = classLevels)
  coefDt = coefDt[geneId != '(Intercept)']

  if (!is.na(geneIdOrder[1])) coefDt = coefDt[geneIdOrder]

  if (ncol(coefDt) == 2) {
    geneSymbols = do.call(c, annotate::lookUp(coefDt$geneId, org, 'SYMBOL', load = TRUE))
    coefDt[, geneId := factor(
      geneId, levels = rev(geneId), labels = sprintf('%s (%s)', rev(geneSymbols), rev(geneId)))]
    p = ggplot(coefDt) + geom_bar(aes(x = .data$geneId, y = .data$coefficient), stat = 'identity')

  } else {
    if (is.na(classLevels[1])) {
      classLevels = colnames(coefDt)[2:ncol(coefDt)]}

    coefDtMolten = melt(coefDt, id.vars = 'geneId', variable.name = 'class',
                        variable.factor = FALSE, value.name = 'coefficient')
    coefDtMolten[, class := factor(class, levels = classLevels)]

    geneIds = coefDt$geneId
    geneSymbols = do.call(c, annotate::lookUp(coefDt$geneId, org, 'SYMBOL', load = TRUE))
    coefDtMolten[, geneId := factor(
      geneId, levels = rev(geneIds), labels = sprintf('%s (%s)', rev(geneSymbols), rev(geneIds)))]

    p = ggplot(coefDtMolten) + facet_wrap(vars(.data$class), ncol = ncol(coefDt) - 1) +
      geom_bar(aes(x = .data$geneId, y = .data$coefficient, fill = .data$class), stat = 'identity') +
      guides(fill = 'none')}

  return(p + coord_flip() + labs(x = 'Gene', y = 'Coefficient') + theme_light())}


#' Plot heatmap of merged gene expression data.
#'
#' Make a heatmap of gene expression from multiple datasets.
#'
#' @param fitResult `glmnet` object.
#' @param lambda value of lambda to use in `fitResult`.
#' @param ematMerged matrix of gene expression for genes by samples.
#' @param sampleMetadata data.frame of sample metadata.
#' @param annoLevels Named list used to make the `annotation`
#'   argument passed to [pheatmap::pheatmap()]. Each name must correspond
#'   to a column name in `sampleMetadata`, and each item in the
#'   list must be a vector of values found in that particular column.
#' @param annoColors Passed to `annotation_colors` argument of
#'   [pheatmap::pheatmap()].
#' @param clusterTogether logical indicating whether to cluster the samples
#'   from each dataset together or separately.
#' @param geneIdOrder Optional character array of Entrez Gene IDs specifying the order
#'   of genes. If `NA` (default), the order from [makeCoefDt()] is used.
#' @param className column in sampleMetadata containing values of the response variable.
#' @param classLevels Order of classes for the column annotations.
#' @param org Name of package for mapping Entrez Gene IDs to gene symbols,
#'   passed to `data` argument of [annotate::lookUp()].
#' @param maxVal Maximum absolute value of scaled and centered gene expression, used
#'   to control dynamic range of color in the heatmap.
#' @param ... Additional arguments passed to [pheatmap::pheatmap()].
#'
#' @return A `pheatmap` object.
#'
#' @export
plotExpressionHeatmap = function(
  fitResult, lambda, ematMerged, sampleMetadata, annoLevels, annoColors = NA,
  clusterTogether = FALSE, geneIdOrder = NA, className = 'class',
  classLevels = NA, org = 'org.Hs.eg', maxVal = 3, ...) {

  geneId = classLevel = NULL
  coefDt = makeCoefDt(fitResult, lambda)
  geneIds = coefDt[geneId != '(Intercept)']$geneId
  geneSymbols = do.call(c, annotate::lookUp(geneIds, org, 'SYMBOL', load = TRUE))
  geneTexts = sprintf('%s (%s)', geneSymbols, geneIds)
  names(geneTexts) = geneIds
  emat = ematMerged[geneIds, ]

  # order the samples
  if (clusterTogether) {
    d = stats::dist(t(emat))
    co = cba::order.optimal(d, stats::hclust(d)$merge)
    emat = emat[, co$order]
  } else {
    if (is.na(classLevels[1])) {
      sm = data.table(sampleMetadata)[sample %in% colnames(ematMerged)]
      classLevels = unique(sm[[className]])}

    emat = foreach(classLevel = classLevels, .combine = cbind) %do% {
      sampleNames1 = sampleMetadata$sample[sampleMetadata[[className]] == classLevel]
      x = emat[, colnames(emat) %in% sampleNames1]
      d = stats::dist(t(x))
      co = cba::order.optimal(d, stats::hclust(d)$merge)
      x = x[, co$order]}}

  # order the genes
  if (is.na(geneIdOrder[1])) {
    d = stats::dist(emat)
    co = cba::order.optimal(d, stats::hclust(d)$merge)
    emat = emat[co$order, ]
    rownames(emat) = geneTexts[co$order]
  } else {
    emat = emat[geneIdOrder, ]
    rownames(emat) = geneTexts[geneIdOrder]}

  # scale the matrix
  emat = t(scale(t(emat)))
  emat[emat > maxVal] = maxVal
  emat[emat < (-maxVal)] = -maxVal

  cols = names(annoLevels)
  annotation = as.data.frame(
    mergeDataTable(colnames(ematMerged), sampleMetadata)[, cols, with = FALSE])
  rownames(annotation) = colnames(ematMerged)

  for (annoName in names(annoLevels)) {
    if (!is.na(annoLevels[[annoName]][1])) {
      annotation[[annoName]] = factor(annotation[[annoName]], levels = annoLevels[[annoName]])}}

  p = pheatmap::pheatmap(
    emat, color = grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = 'RdBu')))(100),
    breaks = seq(-maxVal, maxVal, length.out = 101), cluster_rows = FALSE, cluster_cols = FALSE,
    treeheight_row = 0, treeheight_col = 0, show_colnames = FALSE, border_color = NA,
    annotation_col = annotation, annotation_colors = annoColors, ...)
  return(p)}


#' Plot class probabilities for samples from cross-validation.
#'
#' Make a gtable consisting of multiple `ggplot`s, one for each dataset used in
#' cross-validation. Within each class, samples are sorted by the probability of
#' the true class.
#'
#' @param cvFit `cv.glmnet` object from [metapredictCv()].
#' @param lambda value of lambda to use in `cvFit`.
#' @param ematMerged matrix of gene expression for genes by samples.
#' @param sampleMetadata data.frame of sample metadata.
#' @param className column in sampleMetadata containing values of the response
#'   variable.
#' @param classLevels Preferred order of classes.
#' @param size Point size, passed to [ggplot2::geom_point()].
#' @param ggplotArgs List of additional arguments to add to each ggplot.
#'
#' @return A `ggplot` object constructed by [cowplot::plot_grid()].
#'
#' @export
plotClassProbsCv = function(
  cvFit, lambda, ematMerged, sampleMetadata, className = 'class',
  classLevels = NA, size = 1.5, ggplotArgs = NA) {

  study = trueClass = trueClassProb = idx = NULL
  sampleNames = colnames(ematMerged)
  sm = mergeDataTable(sampleNames, sampleMetadata)
  studyNames = sort(unique(sm$study))

  if (is.na(classLevels[1])) {
    classLevels = sort(unique(sm[[className]]))}

  cvProbs = cvFit$fit.preval[, , which.min(abs(cvFit$lambda - lambda))]
  pList = list()
  for (studyName in studyNames) {
    sampleNamesNow = sm[study == studyName]$sample

    dt = data.table(cvProbs[sm$study == studyName, ])
    setnames(dt, names(cvFit$glmnet.fit$beta))
    dt[, study := studyName]
    dt[, sample := sampleNamesNow]
    dt[, trueClass := factor(sm[sample %in% sampleNamesNow, ][[className]], classLevels)]

    dt[, trueClassProb := apply(dt, MARGIN = 1, function(x) as.numeric(x[x['trueClass']]))]

    setorder(dt, trueClass, -trueClassProb)
    dt = dt[trueClass %in% classLevels]

    idxTmp = c()
    for (classLevel in classLevels) {
      if (any(dt$trueClass == classLevel)) {
        idxTmp = c(idxTmp, 1:(sum(dt$trueClass == classLevel)))}}
    dt[, idx := idxTmp]
    dfMolten = melt(dt, measure.vars = classLevels, variable.name = 'probClass',
                    variable.factor = FALSE, value.name = 'prob')

    p = ggplot(dfMolten) +
      facet_grid(rows = vars(.data$study), cols = vars(.data$trueClass), scales = 'free_x', space = 'free_x') +
      geom_point(aes(x = .data$idx, y = .data$prob, color = .data$probClass, shape = .data$probClass), size = size) +
      labs(x = 'Sample', y = 'Probability') + theme_light() + theme(legend.title = element_blank())

    if (!is.na(ggplotArgs[1])) {
      for (ggplotArg in ggplotArgs) {
        p = p + ggplotArg}}
    pList[[studyName]] = p}

  p = cowplot::plot_grid(plotlist = pList, ncol = 1L, align = 'h')
  return(p)}


#' Plot class probabilities for samples from multiple validation datasets.
#'
#' Make a gtable consisting of multiple ggplots, one for each validation
#' dataset. Within each class, samples are sorted by the probability of the true
#' class.
#'
#' @param predsList Result from [metapredict()].
#' @param sampleMetadata data.frame of sample metadata.
#' @param className column in sampleMetadata containing values of the response
#'   variable.
#' @param classLevels Preferred order of classes.
#' @param size Point size, passed to [ggplot2::geom_point()].
#' @param ggplotArgs List of additional arguments to add to each `ggplot`.
#'
#' @return A ggplot object constructed by [cowplot::plot_grid()].
#'
#' @export
plotClassProbsValidation = function(
  predsList, sampleMetadata, className, classLevels, size = 1.5, ggplotArgs = NA) {

  study = trueClass = trueClassProb = idx = NULL
  pList = list()
  for (validationStudyName in names(predsList)) {
    dt = data.table(predsList[[validationStudyName]][, , 1L], keep.rownames = 'sample')

    sm = mergeDataTable(dt$sample, sampleMetadata)

    dt[, study := sm$study]
    dt[, trueClass := factor(sm[[className]], levels = classLevels)]
    dt[, trueClassProb := apply(dt, MARGIN = 1, function(x) as.numeric(x[x['trueClass']]))]

    setorder(dt, trueClass, -trueClassProb)
    dt = dt[trueClass %in% classLevels]

    idxTmp = c()
    for (classLevel in classLevels) {
      if (any(dt$trueClass == classLevel)) {
        idxTmp = c(idxTmp, 1:(sum(dt$trueClass == classLevel)))}}
    dt[, idx := idxTmp]

    dfMolten = melt(dt, measure.vars = classLevels, variable.name = 'probClass',
                    variable.factor = FALSE, value.name = 'prob')

    p = ggplot(dfMolten) +
      facet_grid(rows = vars(.data$study), cols = vars(.data$trueClass), scales = 'free_x', space = 'free_x') +
      geom_point(aes(x = .data$idx, y = .data$prob, color = .data$probClass, shape = .data$probClass), size = size) +
      labs(x = 'Sample', y = 'Probability') + theme_light() + theme(legend.title = element_blank())

    if (!is.na(ggplotArgs[1])) {
      for (ggplotArg in ggplotArgs) {
        p = p + ggplotArg}}
    pList[[validationStudyName]] = p}

  p = cowplot::plot_grid(plotlist = pList, ncol = 1L, align = 'h')
  return(p)}


R/metapredict_merge.R
makeMatchSampleMapping = function(metadata, subStudyNames, matchSampleColname) {

  study = ..cols = ..matchSampleColname = NULL
  cols = c('study', 'sample')
  if (!(matchSampleColname %in% cols)) cols = c(cols, matchSampleColname)
  metadataNow = data.table(metadata)[study %in% subStudyNames, ..cols]

  order = if (is.unsorted(subStudyNames)) -1L else 1L
  setorderv(metadataNow, 'study', order = order)

  mappingDf = metadataNow[, .SD[1L], by = ..matchSampleColname]
  mapping = mappingDf$sample
  names(mapping) = mappingDf[[matchSampleColname]]
  return(mapping)}


#' Merge gene expression from different platforms that was measured on the same biological samples.
#'
#' @param ematAtomicList list of expression matrices.
#' @param studyMetadataAtomic data frame for study metadata.
#' @param matchStudyColname column in studyMetadataAtomic used to match studies.
#' @param sampleMetadataAtomic data frame for sample metadata.
#' @param matchSampleColname column in sampleMetadataAtomic used to match samples.
#' @param mergeFunc function to summarize multiple gene expression values.
#'
#' @return A named list.
#' \item{ematList}{Named list of expression matrices.}
#' \item{studyMetadata}{data.frame of study metadata.}
#' \item{sampleMetadata}{data.frame of sample metadata.}
#'
#' @export
mergeMatchStudyData = function(
  ematAtomicList, studyMetadataAtomic, matchStudyColname, sampleMetadataAtomic,
  matchSampleColname, mergeFunc = function(x) mean(x, na.rm = TRUE)) {

  study = geneId = ..matchStudyColname = ..colnamesKeep = NULL
  ematList = list()
  sampleMetadataList = list()
  sampleMetadataAtomicDT = data.table(sampleMetadataAtomic)

  for (matchStudyName in unique(studyMetadataAtomic[[matchStudyColname]])) {
    if (sum(studyMetadataAtomic[[matchStudyColname]] == matchStudyName) == 1) {
      ematList[[matchStudyName]] = ematAtomicList[[matchStudyName]]
      sampleMetadataList[[matchStudyName]] = sampleMetadataAtomicDT[study == matchStudyName]

    } else if (sum(studyMetadataAtomic[[matchStudyColname]] == matchStudyName) > 1) {
      atomicStudyNames = studyMetadataAtomic$study[studyMetadataAtomic[[matchStudyColname]] == matchStudyName]
      edfListNow = list()
      for (atomicStudyName in atomicStudyNames) {
        edf = data.frame(rownames(ematAtomicList[[atomicStudyName]]), ematAtomicList[[atomicStudyName]])
        rownames(edf) = NULL

        matchNames = mergeDataTable(colnames(edf)[2:ncol(edf)], sampleMetadataAtomicDT)[[matchSampleColname]]
        colnames(edf) = c('geneId', matchNames)
        edfListNow[[atomicStudyName]] = edf}

      edfMerged = data.frame(data.table(
        rbind(edfListNow)[[1L]])[, lapply(.SD, mergeFunc), by = geneId],
        check.names = FALSE)

      rownames(edfMerged) = edfMerged$geneId
      edfMerged = edfMerged[, -1L]

      mapping = makeMatchSampleMapping(sampleMetadataAtomic, atomicStudyNames, matchSampleColname)
      colnames(edfMerged) = mapping[colnames(edfMerged)]
      ematList[[matchStudyName]] = as.matrix(edfMerged)

      idx = (sampleMetadataAtomic$study %in% atomicStudyNames) &
        (sampleMetadataAtomic$sample %in% colnames(edfMerged))
      sampleMetadataList[[matchStudyName]] = sampleMetadataAtomic[idx, ]
      sampleMetadataList[[matchStudyName]]$study = matchStudyName}}

  colnamesKeep = setdiff(colnames(studyMetadataAtomic), matchStudyColname)

  studyMetadata = studyMetadataAtomic[
    , .SD[1L], by = ..matchStudyColname][
      , study := matchStudyColname][, ..colnamesKeep]

  sampleMetadata = suppressWarnings(rbindlist(sampleMetadataList))

  result = list(ematList, studyMetadata, sampleMetadata)
  names(result) = c('ematList', 'studyMetadata', 'sampleMetadata')
  return(result)}


#' Merge gene expression data from multiple studies.
#'
#' Merge gene expression data from multiple studies, keeping only genes measured
#' in each dataset, then optionally perform cross-study normalization using
#' [sva::ComBat()].
#'
#' @param ematList list of expression matrices.
#' @param sampleMetadata data.frame for sample metadata.
#' @param batchColname column in sampleMetadata containing
#' 	 batch information for ComBat.
#' @param covariateName column in sampleMetadata containing
#' 	 additional covariates for ComBat besides batch.
#' @param batchCorrection `TRUE` indicates cross-study normalization
#' 	 will be performed, `FALSE` indicates datasets will be merged
#' 	 without performing normalization.
#' @param parPrior passed to ComBat's `par.prior` argument.
#'
#' @return A matrix of expression for genes by samples.
#'
#' @export
mergeStudyData = function(
  ematList, sampleMetadata, batchColname = 'study', covariateName = NA,
  batchCorrection = TRUE, parPrior = TRUE) {

  studyName = NULL
  sampleNames = do.call(c, lapply(ematList, colnames))
  if (!all(sampleNames %in% sampleMetadata$sample)) {
    stop('sampleMetadata must have samples corresponding to the colnames of each matrix in ematList.',
         call. = FALSE)}

  ematList = ematList[sapply(ematList, ncol) > 0]
  geneIds = Reduce(intersect, lapply(ematList, function(x) rownames(x)))
  ematList2 = foreach(studyName = names(ematList)) %do% {
    ematNow = ematList[[studyName]][geneIds, ]}

  if (isTRUE(batchCorrection)) {
    # if both one-color and two-color data is present and data is not scaled beforehand,
    # ComBat can fail catastrophically
    ematListScaled = lapply(ematList2, function(emat) (emat - mean(emat)) / stats::sd(emat))
    ematMerged = do.call(cbind, ematListScaled)
    sm = mergeDataTable(colnames(ematMerged), sampleMetadata)

    if (is.na(covariateName) || length(unique(sm[[covariateName]])) < 2) {
      covariateInfo = stats::model.matrix(~ rep_len(1, ncol(ematMerged)))
    } else {
      covariateInfo = stats::model.matrix(~ sm[[covariateName]])}

    if (length(unique(sm[[batchColname]])) > 1) {
      ematMergedNorm = sva::ComBat(
        ematMerged, batch = as.character(sm[[batchColname]]),
        mod = covariateInfo, par.prior = parPrior)
    } else {
      ematMergedNorm = ematMerged}

    return(ematMergedNorm)
  } else {
    return(do.call(cbind, ematList2))}}


R/metapredict_utils.R
#' Install custom CDF packages from Brainarray.
#'
#' Install Brainarray custom CDFs for processing raw Affymetrix data. See
#' <http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/CDF_download.asp>.
#'
#' @param pkgs character vector of package names, e.g., 'hgu133ahsentrezgcdf'
#' @param ver integer version number (25 as of 5 Jan 2021)
#'
#' @export
installCustomCdfPackages = function(pkgs, ver = 25) {
  for (pkg in pkgs) {
    pkgUrl = sprintf(
      'http://mbni.org/customcdf/%d.0.0/entrezg.download/%s_%d.0.0.tar.gz',
      ver, pkg, ver)
    utils::install.packages(pkgUrl, repos = NULL)}}


#' Download custom CDF mapping files from Brainarray.
#'
#' Download Brainarray custom CDF mapping files, which are used for mapping
#' probes to genes in datasets whose `studyDataType` is 'affy_series_matrix'.
#' See
#' \url{http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/CDF_download.asp}.
#'
#' @param cdf `data frame` with columns `download` (e.g., 'Mouse4302_Mm_ENTREZ')
#'   and `rename` (e.g., 'mouse4302mmentrezgcdf').
#' @param path directory into which to download the files.
#' @param ver integer version number (25 as of 5 Jan 2021).
#'
#' @export
downloadCustomCdfMappings = function(cdf, path = '.', ver = 25) {
  if (!dir.exists(path)) {
    dir.create(path)}
  for (ii in seq_len(nrow(cdf))) {
    temp = tempfile()
    utils::download.file(
      sprintf('http://mbni.org/customcdf/%d.0.0/entrezg.download/%s_%d.0.0.zip',
              ver, cdf$download[ii], ver), temp)
    utils::unzip(temp, files = paste0(cdf$download[ii], '_mapping.txt'),
                 exdir = path)
    file.rename(file.path(path, paste0(cdf$download[ii], '_mapping.txt')),
                file.path(path, paste0(cdf$rename[ii], '_mapping.txt')))
    unlink(temp)}}


mergeDataTable = function(sample, sampleMetadata) {
  merge(data.table(sample = sample), sampleMetadata, by = 'sample', sort = FALSE)}


metapredict_example/example_script_orig.R
library('metapredict')
library('ggplot2')
library('RColorBrewer')
library('foreach')
# library('data.table')
# library('dplyr')

if (!(grepl('vignettes', getwd()))) {
  setwd('vignettes')
}

parentFolderPath = 'expression_data'
expressionDataFile = 'expression_data.rds'
denovo = TRUE

foldidColname = 'study'
className = 'class'
family = 'multinomial'
alpha = 0.9
discoveryStudyNames = c('Bhattacharjee', 'GSE11969', 'GSE29016')
classLevels = c('AD', 'SQ', 'SCLC')

rdsPrefix = 'local'

studyMetadataPath = system.file('extdata', 'study_metadata.csv', package = 'metapredict')
studyMetadata = read.csv(studyMetadataPath, stringsAsFactors = FALSE)

sampleMetadataPath = system.file('extdata', 'sample_metadata.csv', package = 'metapredict')
sampleMetadata = read.csv(sampleMetadataPath, stringsAsFactors = FALSE)

foreach(studyRow = iterators::iter(studyMetadata, by = 'row'), .combine = c) %do% {
  if (any(is.na(studyRow))) {
    NA
  } else {
    studyData = getStudyData(parentFolderPath, studyRow$study,
                 studyRow$studyDataType,
                 studyRow$platformInfo)
    saveRDS(studyData, file = paste0(rdsPrefix, studyRow$study, expressionDataFile))}}

if (denovo) {
  esetListOrig = getStudyDataList(parentFolderPath, studyMetadata)
  saveRDS(esetListOrig, file = expressionDataFile)
} else {
  esetListOrig = readRDS(expressionDataFile)}

ematListOrig = extractExpressionData(esetListOrig, sampleMetadata)
ematDiscoveryOrig = mergeStudyData(ematListOrig[discoveryStudyNames], sampleMetadata)

glmnetArgs = makeGlmnetArgs(sampleMetadata[sampleMetadata$study %in% discoveryStudyNames,],
                            foldidColname = foldidColname)

cvFitList = metapredictCv(ematDiscovery, sampleMetadata, yName = className,
                          weights = glmnetArgs$weights, foldid = glmnetArgs$foldid,
                          alpha = alpha, family = family, lambda.min.ratio = 0.001,
                          keep = TRUE)
cvFit = cvFitList[[1]]

plot(cvFit)

fitResult = cvFit$glmnet.fit
lambda = cvFit$lambda.min

plotCoefficients(fitResult, lambda, classLevels = classLevels) +
  scale_fill_brewer(type = 'qual', palette = 3) +
  theme(axis.text.y = element_text(size = 8))

annoNames = c('study', 'class')
annoLevels = list(discoveryStudyNames, classLevels)
names(annoLevels) = annoNames
annoColors = list(brewer.pal(3, 'Dark2'), brewer.pal(3, 'Paired'))
names(annoColors) = annoNames
names(annoColors[[1]]) = annoLevels[[1]]
names(annoColors[[2]]) = annoLevels[[2]]

plotExpressionHeatmap(fitResult, lambda, ematDiscovery, sampleMetadata, annoLevels,
                      annoColors, classLevels = classLevels, fontsize_row = 7)

predsList = metapredict(ematList, studyMetadata, sampleMetadata, discoveryStudyNames,
                        alpha = alpha, lambda = lambda, weights = glmnetArgs$weights,
                        family = family)
#> Found4batches
#> Adjusting for0covariate(s) or covariate level(s)
#> Standardizing Data across genes
#> Fitting L/S model and finding priors
#> Finding parametric adjustments
#> Adjusting the Data

calcConfusionValidation(predsList, lambda = lambda, sampleMetadata = sampleMetadata,
                        classLevels = classLevels)
#> $GSE30219
#>          predictedClass
#> trueClass AD SQ SCLC
#>      AD   85  0    0
#>      SQ    9 52    0
#>      SCLC  4  0   17

plotClassProbsValidation(predsList, sampleMetadata, className, classLevels, size = 1.5,
                         ggplotArgs = list(scale_color_brewer(type = 'qual', palette = 3)))
